drivers/regulator/regulator_axp192.c

```diff
@@ -0,0 +1,355 @@
+
+	ret = i2c_reg_update_byte_dt(&config->i2c, config->desc->enable_reg,
+				     config->desc->enable_mask, 0u);
+	if (ret != 0) {
+		LOG_INST_ERR(config->log, "Failed to disable regulator");
+	}
+
+	return ret;
+}
+
+static unsigned int axp192_count_voltages(const struct device *dev)
+{
+	const struct regulator_axp192_config *config = dev->config;
+
+	return linear_range_group_values_count(config->desc->ranges, config->desc->num_ranges);
+}
+
+static int axp192_list_voltage(const struct device *dev, unsigned int idx, int32_t *volt_uv)
+{
+	const struct regulator_axp192_config *config = dev->config;
+
+	return linear_range_group_get_value(config->desc->ranges, config->desc->num_ranges, idx,
+					    volt_uv);
+}
+
+static int axp192_set_voltage(const struct device *dev, int32_t min_uv, int32_t max_uv)
+{
+	const struct regulator_axp192_config *config = dev->config;
+	uint16_t idx;
+	int ret;
+
+	LOG_INST_DBG(config->log, "voltage = [min=%d, max=%d]", min_uv, max_uv);
+
+	/* set voltage */
+	ret = linear_range_group_get_win_index(config->desc->ranges, config->desc->num_ranges,
+					       min_uv, max_uv, &idx);
+	if (ret != 0) {
+		LOG_INST_ERR(config->log, "No voltage range window could be deteceted");
+		return ret;
+	}
+
+	idx <<= config->desc->vsel_bitpos;
+
+	LOG_INST_DBG(config->log, "[0x%x]=0x%x mask=0x%x", (uint32_t)config->desc->vsel_reg, idx,
+		     (uint32_t)config->desc->vsel_mask);
+	ret = i2c_reg_update_byte_dt(&config->i2c, config->desc->vsel_reg, config->desc->vsel_mask,
+				     (uint8_t)idx);
+	if (ret != 0) {
+		LOG_INST_ERR(config->log, "Failed to set regulator voltage");
+	}
+
+	return ret;
+}
+
+static int axp192_get_voltage(const struct device *dev, int32_t *volt_uv)
+{
+	const struct regulator_axp192_config *config = dev->config;
+	int ret;
+	uint8_t raw_reg;
+
+	/* read voltage */
+	ret = i2c_reg_read_byte_dt(&config->i2c, config->desc->vsel_reg, &raw_reg);
+	if (ret != 0) {
+		return ret;
+	}
+
+	raw_reg = (raw_reg & config->desc->vsel_mask) >> config->desc->vsel_bitpos;
+
+	ret = linear_range_group_get_value(config->desc->ranges, config->desc->num_ranges, raw_reg,
+					   volt_uv);
+
+	return ret;
+}
+
+static int axp192_set_mode(const struct device *dev, regulator_mode_t mode)
+{
+	const struct regulator_axp192_config *config = dev->config;
+	int ret = -ENOTSUP;
+
+	/* setting workmode is only possible for DCDC1-3 */
+	if ((mode == AXP192_DCDC_MODE_PWM) && (config->desc->workmode_reg != 0)) {
+
+		/* configure PWM mode */
+		LOG_INST_DBG(config->log, "PWM mode enabled");
+		ret = i2c_reg_update_byte_dt(&config->i2c, config->desc->workmode_reg,
+					     config->desc->workmode_mask,
+					     config->desc->workmode_pwm_val);
+	} else if (mode == AXP192_DCDC_MODE_AUTO) {
+
+		/* configure AUTO mode (default) */
+		if (config->desc->workmode_reg != 0) {
+			ret = i2c_reg_update_byte_dt(&config->i2c, config->desc->workmode_reg,
+						     config->desc->workmode_mask, 0u);
+		} else {
+
```

Comment:
nit: stray blank line?

Reply:
I generally leave a blank line before a comment, which is more readable to me.